src/compiler/transformers/es6.ts

```diff
@@ -52,20 +184,37 @@ namespace ts {
         function transformSourceFile(node: SourceFile) {
             currentSourceFile = node;
             enclosingBlockScopeContainer = node;
+            currentNode = node;
             return visitEachChild(node, visitor, context);
         }
 
         function visitor(node: Node): VisitResult<Node> {
+            return saveStateAndInvoke(node, dispatcher);
+        }
+
+        function dispatcher(node: Node): VisitResult<Node> {
+            return convertedLoopState
+                ? visitorForConvertedLoopWorker(node)
+                : visitorWorker(node);
+        }
+
+        function saveStateAndInvoke<T>(node: Node, f: (node: Node) => T): T {
             const savedContainingNonArrowFunction = containingNonArrowFunction;
             const savedCurrentParent = currentParent;
             const savedCurrentNode = currentNode;
             const savedEnclosingBlockScopeContainer = enclosingBlockScopeContainer;
             const savedEnclosingBlockScopeContainerParent = enclosingBlockScopeContainerParent;
 
-            onBeforeVisitNode(node);
+            const savedConvertedLoopState = convertedLoopState;
+            if (nodeStartsNewLexicalEnvironment(node)) {
+                // don't treat content of nodes that start new lexical environment or class-like nodes as part of converted loop copy 

```

Comment:
This comment may now be incorrect.

How do we want to handle captured block scope bindings for initializers?

```ts
let ar = [];
for (let i = 0; i < 2; i++) {
  ar.push(class { x = i; });
}
console.log(new ar[0]().x); // should this be 0 or 2?
console.log(new ar[1]().x); // should this be 1 or 2?
```